### PR TITLE
Add diff in Groupby

### PR DESCRIPTION
**Pull request by RainFung**
This PR adds rank GroupBy (both SeriesGroupBy and DataFrameGroupBy) by using existing `diff` logic in Series.

### CODE REVIEW

File: databricks/koalas/groupby.py [hunk 1]
```diff
@@ -418,6 +419,81 @@ def size(self):
...
+        2    NaN
+        3    NaN
+        4    NaN
+        5    NaN
+        Name: a, dtype: float64
+        """
+        sdf = self._kdf._sdf
+        index = self._kdf._internal.index_columns[0]
+        for i, s in enumerate(self._agg_columns):
+            window = Window.partitionBy([s._scol for s in self._groupkeys])\
```
Comment (HyukjinKwon) [inline databricks/koalas/groupby.py]: This one too. Can we deduplicate with `Series.diff`?

File: databricks/koalas/groupby.py [hunk 2]
```diff
@@ -422,6 +423,81 @@ def size(self):
...
         return _col(DataFrame(internal))
 
+    def diff(self, periods=1):
+        """
+        First discrete difference of element.
+
+        Calculates the difference of a DataFrame element compared with another element in the
+        DataFrame (default is the element in the same column of the previous row).
+
+        .. note:: the current implementation of diff uses Spark's Window without
```
Comment (HyukjinKwon) [inline databricks/koalas/groupby.py]: If we use partition columns, the data will be distributed by its partition columns. So we could remove this note.

File: databricks/koalas/groupby.py [hunk 3]
```diff
@@ -422,6 +423,81 @@ def size(self):
...
+        Name: a, dtype: float64
+        """
+        sdf = self._kdf._sdf
+        index = self._kdf._internal.index_columns[0]
+        for i, s in enumerate(self._agg_columns):
+            window = Window.partitionBy([s._scol for s in self._groupkeys])\
+                .orderBy(index).rowsBetween(-periods, -periods)
+            sdf = sdf.withColumn(s.name, s._scol - F.lag(s._scol, periods).over(window))
+
+        internal = _InternalFrame(sdf=sdf.orderBy(index),
```
Comment (HyukjinKwon) [inline databricks/koalas/groupby.py]: Let's don't bother to sort by index for now.

**Comment by HyukjinKwon** [comment]
Looks fine in general.

File: databricks/koalas/groupby.py (hunk 1 above)
Comment (RainFung) [reply databricks/koalas/groupby.py]: I think there is no need to deduplicate with `Series.diff`.Unlike `Series.diff`, there is `partitionBy ` in `groupby.diff`.

File: databricks/koalas/groupby.py (hunk 2 above)
Comment (HyukjinKwon) [reply databricks/koalas/groupby.py]: @RainFung, looks this comment and the comment below were not addressed.

File: databricks/koalas/groupby.py (hunk 1 above)
Comment (HyukjinKwon) [reply databricks/koalas/groupby.py]: you can do it like https://github.com/databricks/koalas/commit/429e26be5bbbf90f7887cfc5531ec1210a7f5f11 or https://github.com/databricks/koalas/commit/510cad590461e5ea42fb2e2a4846f62ae55a210f

**Comment by HyukjinKwon** [comment]
Thanks, @RainFung for working on this.

**Comment by ppakawatk** [comment]
Hi,

I'm trying to convert my Pandas script to Koalas script, and found this error when trying to do GroupBy.diff with datetime column.

Here's the code to reproduce the case:
`df = ks.DataFrame({'a': [1, 2, 3, 4, 5, 6], 'b': [1, 1, 2, 3, 5, 8], 'c': [dt.date(2019, 1, 1), dt.date(2019, 1, 3), dt.date(2019, 1, 10), dt.date(2019, 1, 11), dt.date(2019, 1, 20), dt.date(2019, 1, 21)]}, columns=['a', 'b', 'c'])
df.groupby(['b'])['c'].diff()`

Referring to https://koalas.readthedocs.io/en/latest/reference/api/databricks.koalas.groupby.GroupBy.diff.html
I assume that right now it doesn't support datetime column when diff? 


PS. here is the Pandas script which is working properly.
`df = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6], 'b': [1, 1, 2, 3, 5, 8], 'c': [dt.date(2019, 1, 1), dt.date(2019, 1, 3), dt.date(2019, 1, 10), dt.date(2019, 1, 11), dt.date(2019, 1, 20), dt.date(2019, 1, 21)]}, columns=['a', 'b', 'c'])
df.groupby(['b'])['c'].diff()`

**Comment by HyukjinKwon** [comment]
Can you open an issue with runnable examples? Seems like it's also related with Calender internal type support (SPARK-24695)

**Comment by ppakawatk** [comment]
> Can you open an issue with runnable examples? Seems like it's also related with Calender internal type support (SPARK-24695)

Yes, sir.
https://github.com/databricks/koalas/issues/965

**Pull request merged**